File: mobile/android/app/src/main/java/org/modelbot/mobile/Workspace.java
package org.modelbot.mobile;

import org.json.JSONObject;
import java.io.InputStream;
import java.net.HttpCookie;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URL;
import java.nio.charset.StandardCharsets;

final class Workspace {
    volatile String origin = "", cookie = "", csrf = "";
    private int generation;
    static final class Expired extends Exception { Expired() { super("Sign-in expired. Paste a fresh workspace connection link. Tasks remain on the server."); } }
    final SessionVault vault;
    Workspace(SessionVault vault) { this.vault = vault; }
    static URI connectionURL(String input) throws Exception {
        URI url; try { url = new URI(input.trim()); } catch (java.net.URISyntaxException error) { throw new Exception("The connection link is invalid. Copy a fresh link from your workspace."); } String host = url.getHost();
        boolean local = host != null && (host.equals("localhost") || host.equals("127.0.0.1"));
        if (host == null || url.getUserInfo() != null || url.getRawQuery() != null || !(url.getPath().isEmpty() || url.getPath().equals("/"))
            || !("https".equalsIgnoreCase(url.getScheme()) || (BuildConfig.DEBUG && local && "http".equalsIgnoreCase(url.getScheme()))))
            throw new Exception("Use your workspace’s HTTPS connection link. HTTP is allowed only for local debug tests.");
        return url;
    }
    void restore() throws Exception {
        String saved = vault.read(); if (saved == null) return;
        JSONObject data = new JSONObject(saved); origin = data.getString("origin"); connectionURL(origin); cookie = data.getString("cookie");
    }
    JSONObject connect(String input) throws Exception {
        URI url = connectionURL(input);
        final int version; synchronized(this) { version = ++generation; }
        String next = new URI(url.getScheme().toLowerCase(java.util.Locale.ROOT), null, url.getHost().toLowerCase(java.util.Locale.ROOT), ("https".equalsIgnoreCase(url.getScheme()) && url.getPort()==443) ? -1 : url.getPort(), null, null, null).toString();
        String token = null;
        if (url.getRawFragment() != null) for (String pair : url.getRawFragment().split("&")) {
            String[] entry = pair.split("=", 2);
            if (entry.length == 2 && entry[0].equals("bootstrap")) token = java.net.URLDecoder.decode(entry[1], "UTF-8");
        }
        if (!next.equals(origin)) { cookie = ""; csrf = ""; }
        origin = next;
        JSONObject session = token == null ? request("GET", "/api/v1/session", null) : request("POST", "/api/v1/session/bootstrap", new JSONObject().put("token", token).put("label", "Android"));
        csrf = session.optString("csrf");
        if (cookie.isEmpty() || csrf.isEmpty()) throw new Exception("Paste a fresh connection link from your workspace to sign in.");
        synchronized(this) { if(version != generation) throw new Exception("The connection changed. Open the current workspace again."); vault.save(new JSONObject().put("origin", origin).put("cookie", cookie).toString()); }
        return session;
    }
    JSONObject request(String method, String path, JSONObject body) throws Exception {
        final int version; final String base, sessionCookie, requestCsrf;
        synchronized(this) { version = generation; base = origin; sessionCookie = cookie; requestCsrf = csrf; }
        if (!path.startsWith("/api/v1/")) throw new Exception("Invalid workspace request.");
        HttpURLConnection connection = (HttpURLConnection) new URL(base + path).openConnection();
        connection.setInstanceFollowRedirects(false); connection.setConnectTimeout(15000); connection.setReadTimeout(30000);
        connection.setRequestMethod(method); connection.setRequestProperty("Origin", base); connection.setRequestProperty("Accept", "application/json");
        if (!sessionCookie.isEmpty()) connection.setRequestProperty("Cookie", sessionCookie);
        if (!requestCsrf.isEmpty()) connection.setRequestProperty("X-CSRF-Token", requestCsrf);
        try {
            if (body != null) { connection.setDoOutput(true); connection.setRequestProperty("Content-Type", "application/json"); connection.getOutputStream().write(body.toString().getBytes(StandardCharsets.UTF_8)); }
            int code = connection.getResponseCode();
            synchronized(this) { if(version != generation) throw new Exception("The connection changed. Open the current workspace again."); if (code == 401 || code == 403) { generation++; cookie=""; csrf=""; vault.clear(); throw new Expired(); } }
            if (code < 200 || code >= 300) throw new Exception("The workspace could not complete this request (" + code + "). Check its status before trying again.");
            String setCookie = connection.getHeaderField("Set-Cookie");
            synchronized(this) { if(version != generation) throw new Exception("The connection changed."); if (setCookie != null) for (HttpCookie value : HttpCookie.parse(setCookie)) if (value.getName().equals("modelbot_session")) cookie = value.getName() + "=" + value.getValue(); }
            try (InputStream stream = connection.getInputStream()) {
                java.io.ByteArrayOutputStream buffer = new java.io.ByteArrayOutputStream(); byte[] chunk = new byte[8192]; int count;
                while ((count = stream.read(chunk)) != -1) { if (buffer.size() + count > 2_000_000) throw new Exception("Workspace response exceeds this app’s size limit."); buffer.write(chunk, 0, count); }
                byte[] bytes = buffer.toByteArray();
                JSONObject result = new JSONObject(new String(bytes, StandardCharsets.UTF_8));
                synchronized(this) { if(version != generation) throw new Exception("The connection changed."); if (path.equals("/api/v1/session")) csrf = result.optString("csrf"); }
                return result;
            }
        } finally { connection.disconnect(); }
    }
    synchronized void forget() { generation++; origin = ""; cookie = ""; csrf = ""; vault.clear(); }
}

File: mobile/ios/ModelBot/Workspace.swift
import Foundation
import Security

struct TaskItem: Decodable, Identifiable {
  let id: String
  let computer_id: String
  let goal: String
  let status: String
}
struct Takeover: Decodable, Identifiable {
  let id: String
  let computer_id: String
  let state: String
}
struct SavedSession: Codable {
  let origin: String
  let cookie: String
}

enum SessionVault {
  private static let query: [String: Any] = [
    kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: "ModelBot.operator",
    kSecAttrAccount as String: "workspace",
  ]
  static func read() throws -> SavedSession? {
    var query = query
    query[kSecReturnData as String] = true
    query[kSecMatchLimit as String] = kSecMatchLimitOne
    var value: CFTypeRef?
    let status = SecItemCopyMatching(query as CFDictionary, &value)
    if status == errSecItemNotFound { return nil }
    guard status == errSecSuccess, let data = value as? Data else {
      throw WorkspaceError.message("Unlock this device to restore your connection.")
    }
    return try JSONDecoder().decode(SavedSession.self, from: data)
  }
  static func save(_ value: SavedSession) throws {
    let data = try JSONEncoder().encode(value)
    let status = SecItemUpdate(
      query as CFDictionary, [kSecValueData as String: data] as CFDictionary)
    if status == errSecItemNotFound {
      var item = query
      item[kSecValueData as String] = data
      item[kSecAttrAccessible as String] = kSecAttrAccessibleWhenUnlockedThisDeviceOnly
      guard SecItemAdd(item as CFDictionary, nil) == errSecSuccess else {
        throw WorkspaceError.message("The connection could not be saved securely.")
      }
      return
    }
    guard status == errSecSuccess else {
      throw WorkspaceError.message("The connection could not be saved securely.")
    }
  }
  static func clear() { SecItemDelete(query as CFDictionary) }
}
enum WorkspaceError: LocalizedError {
  case message(String)
  case expired
  var errorDescription: String? {
    switch self {
    case .message(let text): return text
    case .expired:
      return "Sign-in expired. Paste a fresh workspace connection link; tasks remain on the server."
    }
  }
}

/// Same-origin operator HTTP client; provider credentials never enter this app.
@MainActor final class WorkspaceClient: NSObject, URLSessionTaskDelegate {
  var origin = "", cookie = "", csrf = ""
  private var generation = 0
  private lazy var session: URLSession = {
    let config = URLSessionConfiguration.ephemeral
    config.httpShouldSetCookies = false
    config.httpCookieStorage = nil
    config.timeoutIntervalForRequest = 30
    return URLSession(configuration: config, delegate: self, delegateQueue: nil)
  }()
  static func connectionURL(_ input: String) throws -> URLComponents {
    guard let url = URLComponents(string: input.trimmingCharacters(in: .whitespacesAndNewlines)),
      let host = url.host,
      url.user == nil, url.password == nil, url.query == nil, url.path.isEmpty || url.path == "/"
    else {
      throw WorkspaceError.message("Use the HTTPS connection link from your own workspace.")
    }
    var allowed = url.scheme?.lowercased() == "https"
    #if DEBUG
      allowed =
        allowed
        || (url.scheme?.lowercased() == "http" && ["localhost", "127.0.0.1", "::1"].contains(host))
    #endif
    guard allowed else {
      throw WorkspaceError.message(
        "Your workspace needs HTTPS. Insecure remote connections are not supported.")
    }
    return url
  }
  func restore() throws {
    guard let saved = try SessionVault.read() else { return }
    _ = try Self.connectionURL(saved.origin)
    origin = saved.origin
    cookie = saved.cookie
  }
  func connect(_ input: String) async throws -> [String: Any] {
    generation += 1
    var url = try Self.connectionURL(input)
    let fragment = URLComponents(string: "?" + (url.fragment ?? ""))?.queryItems
    let token = fragment?.first(where: { $0.name == "bootstrap" })?.value
    url.fragment = nil
    url.path = ""
    url.scheme = url.scheme?.lowercased()
    url.host = url.host?.lowercased()
    if url.scheme?.lowercased() == "https" && url.port == 443 { url.port = nil }
    guard let next = url.string else {
      throw WorkspaceError.message("The connection link is invalid.")
    }
    if next != origin {
      cookie = ""
      csrf = ""
    }
    origin = next
    let result = try await request(
      token == nil ? "GET" : "POST", token == nil ? "/api/v1/session" : "/api/v1/session/bootstrap",
      token.map { ["token": $0, "label": "iPhone / iPad"] })
    csrf = result["csrf"] as? String ?? ""
    guard !cookie.isEmpty, !csrf.isEmpty else {
      throw WorkspaceError.message("Paste a fresh workspace connection link to sign in.")
    }
    try SessionVault.save(SavedSession(origin: origin, cookie: cookie))
    return result
  }
  func request(_ method: String, _ path: String, _ body: [String: Any]? = nil) async throws
    -> [String: Any]
  {
    let currentGeneration = generation
    guard path.hasPrefix("/api/v1/"), let url = URL(string: origin + path) else {
      throw WorkspaceError.message("Connect a workspace first.")
    }
    var request = URLRequest(url: url)
    request.httpMethod = method
    request.setValue(origin, forHTTPHeaderField: "Origin")
    request.setValue("application/json", forHTTPHeaderField: "Accept")
    if !cookie.isEmpty { request.setValue(cookie, forHTTPHeaderField: "Cookie") }
    if !csrf.isEmpty { request.setValue(csrf, forHTTPHeaderField: "X-CSRF-Token") }
    if let body {
      request.httpBody = try JSONSerialization.data(withJSONObject: body)
      request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    }
    let (data, response) = try await session.data(for: request)
    guard generation == currentGeneration else {
      throw WorkspaceError.message("The connection changed. Open the current workspace again.")
    }
    guard let response = response as? HTTPURLResponse else {
      throw WorkspaceError.message("The workspace could not be reached.")
    }
    if [401, 403].contains(response.statusCode) {
      generation += 1
      cookie = ""
      csrf = ""
      SessionVault.clear()
      throw WorkspaceError.expired
    }
    guard (200...299).contains(response.statusCode) else {
      throw WorkspaceError.message(
        "The workspace could not complete this request (\(response.statusCode)). Check its status before trying again."
      )
    }
    guard data.count <= 2_000_000 else {
      throw WorkspaceError.message("The response exceeds this app’s size limit.")
    }
    if let fields = response.allHeaderFields as? [String: String],
      let received = HTTPCookie.cookies(withResponseHeaderFields: fields, for: url).first(where: {
        $0.name.range(of: #"^modelbot_session(?:_[0-9]+)?$"#, options: .regularExpression) != nil
      })
    {
      cookie = "\(received.name)=\(received.value)"
    }
    guard let result = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
      throw WorkspaceError.message("The workspace response could not be read.")
    }
    if path == "/api/v1/session" { csrf = result["csrf"] as? String ?? "" }
    return result
  }
  func forget() {
    generation += 1
    origin = ""
    cookie = ""
    csrf = ""
    SessionVault.clear()
  }
  nonisolated func urlSession(
    _ session: URLSession, task: URLSessionTask,
    willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest,
    completionHandler: @escaping (URLRequest?) -> Void
  ) { completionHandler(nil) }
}
